src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+						apps[appId].services[service.imageId].status = 'Handing over'
+
+				_.forEach images, (image) ->
+					appId = image.appId
+					if !image.dependent
+						apps[appId] ?= {}
+						apps[appId].services ?= {}
+						apps[appId].services[image.imageId] ?= _.pick(image, [ 'status', 'download_progress', 'releaseId' ])
+					else
+						dependent[appId] ?= {}
+						dependent[appId].images ?= {}
+						dependent[appId].images[image.imageId] = _.pick(image, [ 'status', 'download_progress' ])
+
+				obj = { local: apps, dependent }
+				obj.commit = commit if commit
+				return obj
+		)
+
+	getDependentState: =>
+		@proxyvisor.getCurrentStates()
+
+	_buildApps: (services, networks, volumes) ->
+		apps = _.keyBy(_.map(_.uniq(_.map(services, 'appId')), (appId) -> { appId }), 'appId')
+
+		# We iterate over the current running services and add them to the current state
+		# of the app they belong to.
+		_.forEach services, (service) ->
+			appId = service.appId
+			apps[appId].services ?= []
+			apps[appId].services.push(service)
+
+		_.forEach networks, (network) ->
+			appId = network.appId
+			apps[appId] ?= { appId }
+			apps[appId].networks ?= {}
+			apps[appId].networks[network.name] = network.config
+
+		_.forEach volumes, (volume) ->
+			appId = volume.appId
+			apps[appId] ?= { appId }
+			apps[appId].volumes ?= {}
+			apps[appId].volumes[volume.name] = volume.config
+
+		# We return the apps as an array
+		return _.values(apps)
+
+	getCurrentForComparison: =>
+		Promise.join(
+			@services.getAll()
+			@networks.getAll()
+			@volumes.getAll()
+			(services, networks, volumes) =>
+				return @_buildApps(services, networks, volumes)
+		)
+
+	getCurrentApp: (appId) =>
+		Promise.join(
+			@services.getAllByAppId(appId)
+			@networks.getAllByAppId(appId)
+			@volumes.getAllByAppId(appId)
+			(services, networks, volumes) =>
+				return @_buildApps(services, networks, volumes)[0]
+		)
+
+	getTargetApp: (appId) =>
+		@db.models('app').where({ appId }).select()
+		.then ([ app ]) =>
+			return if !app?
+			@normaliseAndExtendAppFromDB(app)
+
+	# Compares current and target services and returns a list of service pairs to be updated/removed/installed.
+	# The returned list is an array of objects where the "current" and "target" properties define the update pair, and either can be null
+	# (in the case of an install or removal).
+	compareServicesForUpdate: (currentServices, targetServices) ->
+		Promise.try ->
+			removePairs = []
+			installPairs = []
+			updatePairs = []
+			targetServiceIds = _.map(targetServices, 'serviceId')
+			currentServiceIds = _.uniq(_.map(currentServices, 'serviceId'))
+
+			toBeRemoved = _.difference(currentServiceIds, targetServiceIds)
+			_.forEach toBeRemoved, (serviceId) ->
+				servicesToRemove = _.filter(currentServices, (s) -> s.serviceId == serviceId)
+				_.map servicesToRemove, (service) ->
+					removePairs.push({
+						current: service
+						target: null
```

Comment:
Why do you pass `target: null` here since the collection only contains remove operations?

Reply:
For consistency, I want all the collections returned by this method to have current, target and serviceId.